Remove nontransitive from functions, add mypy type hints, fix linter issues

kinow opened the pull request:
Closes #2570
Closes #2551
Moves some code from #2474 (under review) to simplify that PR

- mark nontransitive and no_recover_logs deprecated in autosubmit monitor
- create grouping only when plotting
- fixes IDE and linter issues
- adds mypy type hints
- remove unused code

<!--

Thanks for your contribution! Please:
* List any related issues with a "closes" or "addresses" tag.
* Add a helpful title & description.
* Complete the checklist.

-->

**Check List**

- [x] I have read `CONTRIBUTING.md`.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Applied any dependency changes to `pyproject.toml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included in `CHANGELOG.md` if this is a change that can affect users.
- [ ] Documentation updated. (no docs needed)
- [x] If this is a bug fix, PR should include a link to the issue (e.g. `Closes #1234`).

kinow commented:
I think Slurm tests are the only ones failing. Will work on the coverage, then ask @VindeeR to have a look at this.

kinow commented:
Huh, `test-slurm` passing on GH Actions. Interesting, I'm getting a bunch of connection time out errors when running locally.

kinow commented on autosubmit/job/job.py:
```diff
@@ -1621,8 +1601,7 @@ def update_current_parameters(self, as_conf: AutosubmitConfig, parameters: dict)
         if not self.platform:
-            submitter = job_utils._get_submitter(as_conf)
-            submitter.load_platforms(as_conf)
+            submitter = ParamikoSubmitter(as_conf=as_conf)
```
Now we simply declare `ParamikoSubmitter` and pass the `as_conf` object -- when the object is passed, the constructor of the class will call `self.load_platforms`, simplifying a bit.

kinow commented on test/integration/platforms/test_slurmplatform.py:
```diff
@@ -359,4 +380,4 @@ def test_run_all_wrappers_workflow_slurm(autosubmit_exp: AutosubmitExperimentFix
 
     exp.autosubmit._check_ownership_and_set_last_command(exp.as_conf, exp.expid, 'run')
-    assert 0 == exp.autosubmit.run_experiment(_EXPID)
+    assert 0 == exp.autosubmit.run_experiment(exp.expid)
```
@VindeeR , ignore the rest here for now. This file contains some changes that I expect will make the test_slurm stop failing in this branch :+1: You may copy to one of your other branches to confirm it fixes the intermittent failures.

kinow commented on autosubmit/platforms/paramiko_submitter.py:
```diff
@@ -83,31 +97,18 @@ def load_platforms(self, asconf, retries=5, auth_password = None, local_auth_pas
-        local_platform.host = 'localhost'
-        # Add object to entry in dictionary
-        platforms['LOCAL'] = local_platform
+        self.load_local_platform(as_conf=as_conf, exp_data=exp_data)
```
This code was almost an identical copy of another part of the code in this same file. This new function combines the behaviour of both to reduce code duplication.

kinow commented on autosubmit/autosubmit.py:
```diff
@@ -727,35 +731,40 @@ def parse_args() -> Tuple[int, Optional[argparse.Namespace]]:
         expid = "None"
+
+        if hasattr(args, 'notransitive') and args.notransitive:
+            warnings.warn('notransitive is deprecated and will be removed in a future major release!')
```
I used `warnings` to display a deprecation warning if any user tries to pass `--notransitive` :+1: There are some other args that I removed but that are not so general, so the changes are within the `if` blocks below for those commands.

kinow commented:
I'm kicking the slurm build.

kinow commented:
@VindeeR I noticed you have `paramiko_platform.py` in this PR. I'm trying to cover as much as I can in the PR https://github.com/BSC-ES/autosubmit/pull/2604

If you want you can skip that class for now. I hope to finish all tests this week, and maybe if that's merged first, then you can rebase yours and maybe it will be fully covered :)

VindeeR commented:
@kinow I tried to add your name to do the review but I learned that I can't :rofl: 
Anyway, I finished the tests so once you're free could you please review this :smiley:

kinow commented:
All good @VindeeR . Once we merge the other PR with tests for the platforms I will rebase this one and review it. Cheers

kinow reviewed (commented):
Did you modify `submitter.py` here? I think it's showing here because you have one merge commit somewhere. Maybe if you squash your commits then it will display only what was modified.

kinow commented:
> > Did you modify `submitter.py` here? I think it's showing here because you have one merge commit somewhere. Maybe if you squash your commits then it will display only what was modified.
> 
> you removed the submitter iirc when you did the change here, should I keep it?

Oh, d'oh. I thought I was going to do that in another PR. :facepalm: I think I need to rest :mask: sorry Erick!

kinow commented:
Rebased. Seeing where I stopped on this now...

VindeeR commented:
@kinow I believe once we finish this branch we going to be over 70% coverage :raised_hands: LESGOOO

kinow commented:
The first batch of failures was easy to fix, just a rebase accident where I removed an argument. Let me try this last one. It's the same that was failing last week when I stopped working on this one.

```
 =========================== short test summary info ============================
FAILED test/integration/git/test_autosubmit_git.py::test_git_operational_experiment_toggle_flag[False-0] - autosubmit.log.log.AutosubmitCritical:
===== 1 failed, 246 passed, 17 skipped, 7607 warnings in 142.63s (0:02:22) =====
```

kinow commented:
Snap. Still failing. But at least I think I fixed all other tests. Will try to fix this one tomorrow morning. Will do some email reading, admin work, etc, to see if I can think of other ways to fix this issue. I think you can start reviewing the PR, @VindeeR , marking files as Read/Reviewed in the GitHub UI. Then tomorrow or Wednesday I hope to have all tests passing, or we may have to ignore the test_send_data for now.

VindeeR commented on autosubmit/job/job_list.py:
```diff
@@ -930,11 +931,11 @@ def _filter_current_job(self, current_job, relationships):
-    def _add_edges_map_info(self, job, special_status):
+    def _add_edges_map_info(self, job: Job, special_status: str):
         """
         Special relations to be check in the update_list method
```
Adjust Docstring

VindeeR commented on autosubmit/job/job_list.py:
```diff
@@ -2973,7 +2972,7 @@ def save_wrappers(self, packages_to_save, failed_packages, as_conf, packages_per
+    def check_scripts(self, as_conf) -> bool:
         """
         When we have created the scripts, all parameters should have been substituted.
         %PARAMETER% handlers not allowed
```
Adjust Docstring

VindeeR commented on autosubmit/migrate/migrate.py:
```diff
@@ -161,11 +162,12 @@ def migrate_pickup(self):
-    def check_migrate_config(self, as_conf, platforms_to_test, pickup_data ):
+    def check_migrate_config(self, as_conf, platforms_to_test, pickup_data):
         """
         Checks if the configuration file has the necessary information to migrate the data
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/headers/slurm_header.py:
```diff
@@ -70,12 +77,13 @@ def get_proccesors_directive(self, job, parameters, het=-1):
+    def get_partition_directive(self, job: 'Job', parameters, het: int = -1) -> str:
         """
         Returns partition directive for the specified job
 
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/headers/slurm_header.py:
```diff
@@ -122,11 +135,13 @@ def get_exclusive_directive(self, job, parameters, het=-1):
+    def get_nodes_directive(self, job: 'Job', parameters, het=-1) -> str:
         """
         Returns nodes directive for the specified job
         :param job: job to create nodes directive for
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/headers/slurm_header.py:
```diff
@@ -139,12 +154,14 @@ def get_nodes_directive(self, job, parameters, het=-1):
+    def get_memory_directive(self, job: 'Job', parameters, het=-1):
         """
         Returns memory directive for the specified job
 
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/locplatform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/paramiko_platform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/paramiko_platform.py:
```diff
@@ -1333,7 +1354,7 @@ def parse_job_output(self, output):
-    def parse_Alljobs_output(self, output, job_id):
+    def parse_all_jobs_output(self, output, job_id):
         """
         Parses check jobs command output, so it can be interpreted by autosubmit
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/paramiko_platform.py:
```diff
@@ -1351,7 +1372,7 @@ def generate_submit_script(self):
-    def get_submit_cmd(self, job_script, job, hold=False, export=""):
+    def get_submit_cmd(self, job_script: str, job, hold: bool = False, export: str = "") -> str:
         """
         Get command to add job to scheduler
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/pjmplatform.py:
```diff
@@ -79,17 +79,15 @@ def generate_new_name_submit_script_file(self):
-    def process_batch_ready_jobs(self,valid_packages_to_submit,failed_packages,error_message="",hold=False):
+    def process_batch_ready_jobs(self, valid_packages_to_submit, failed_packages, error_message="", hold=False):
         """
         Retrieve multiple jobs identifiers.
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/pjmplatform.py:
```diff
@@ -335,7 +322,7 @@ def parse_Alljobs_output(self, output, job_id):
-    def parse_joblist(self, job_list: list[list['Job']]) -> str:
+    def parse_job_list(self, job_list: list[list['Job']]) -> str:
         """
         Convert a list of job_list to job_list_cmd
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/platform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/platform.py:
```diff
@@ -670,16 +665,14 @@ def get_logs_files(self, exp_id: str, remote_logs: tuple[str, str]) -> None:
     def get_checkpoint_files(self, job):
         """
         Get all the checkpoint files of a job
 
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/platform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
```diff
@@ -464,69 +449,37 @@ def parse_job_output(self, output: str) -> str:
-        :rtype: Union[list[str], str]
-        """
+    def parse_all_jobs_output(self, output: str, job_id: int) -> Union[list[str], str]:
         status = ""
```
Rollback Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
```diff
@@ -464,69 +449,37 @@ def parse_job_output(self, output: str) -> str:
-        :rtype: Union[list[int], int]
-        """
+    def get_submitted_job_id(self, output: str, x11: bool = False) -> Union[list[int], int]:
         try:
```
Rollback and adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
```diff
@@ -551,25 +504,19 @@ def get_submit_cmd(self, job_script: str, job, hold: bool = False, export: str =
+    def get_check_job_cmd(self, job_id: str) -> str:
         """
         Generates sacct command to the job selected.
 
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
```diff
@@ -690,13 +625,12 @@ def get_queue_status(self, in_queue_jobs: List['Job'], list_queue_jobid: str, as
+    def wrapper_header(self, **kwargs: Any) -> str:
         """
         It generates the header of the wrapper configuring it to execute the Experiment.
 
```
Adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
Adjust Docstring

VindeeR commented on autosubmit/platforms/slurmplatform.py:
Adjust Docstring

VindeeR commented on autosubmit/autosubmit.py:
```diff
@@ -2264,11 +2295,11 @@ def refresh_log_recovery_process(platforms: list[Platform], as_conf: AutosubmitC
+    def run_experiment(expid: str, start_time: Optional[str] = None, start_after: Optional[str] = None,
+                       run_only_members: Optional[str] = None, profile=False) -> int:
         """
         Runs and experiment (submitting all the jobs properly and repeating its execution in case of failure).
```
Adjust Docstring

VindeeR commented on autosubmit/autosubmit.py:
```diff
@@ -5696,43 +5726,30 @@ def _get_job_list_persistence(expid, as_conf) -> JobListPersistence:
+                 minimal_configuration=False, git_repo=None, git_branch=None, git_as_conf=None,
+                 use_local_minimal=False):
         """
         Method to conduct a test for a given experiment. It creates a new experiment for a given experiment with a
```
Adjust Docstring

VindeeR commented on test/integration/test_expid.py:
```diff
@@ -291,12 +291,12 @@ def _get_experiment_data(tmp_path):
-                'scratch': str(tmp_path / 'scratch'),
+            'PYTEST-PS': {
+                'TYPE': 'ps',
+                'HOST': 'localDocker',
```
Rollback to `127.0.0.1`

kinow commented:
@VindeeR 's last commit from yesterday passed the build. I squashed the commit and rebased, and the build passed again with an entry to the changelog, and the build passed again :tada: I **hope** the fix of using separated directories has fixed the `test_send_data` (again!).

@VindeeR I'm +1 for this, can you have a final look and re-approve it, please? If that's approved I'll merge it today to rebase the operational issue. Thanks!

kinow commented:
Thanks @VindeeR !
